nearcore/src/shard_tracker.rs

```diff
@@ -1,95 +1,90 @@
-use std::collections::{HashMap, HashSet};
+use std::collections::HashMap;
 use std::sync::{Arc, RwLock};
 
-use tracing::info;
-
 use near_epoch_manager::EpochManager;
+use near_primitives::errors::EpochError;
 use near_primitives::hash::CryptoHash;
-use near_primitives::shard_layout::{account_id_to_shard_id, ShardLayout};
+use near_primitives::shard_layout::account_id_to_shard_id;
 use near_primitives::types::{AccountId, EpochId, ShardId};
+use std::hash::Hash;
 
 const POISONED_LOCK_ERR: &str = "The lock was poisoned.";
 
+struct AppendOnlyMap<K, V> {
+    map: RwLock<HashMap<K, Arc<V>>>,
+}
+
+impl<K, V> AppendOnlyMap<K, V>
+where
+    K: Eq + Hash + Clone,
+{
+    fn new() -> Self {
+        Self { map: RwLock::new(HashMap::new()) }
+    }
+
+    fn get_or_insert<F: FnOnce() -> V>(&self, key: &K, value: F) -> Arc<V> {
+        let mut map = self.map.write().expect(POISONED_LOCK_ERR);
+        if !map.contains_key(key) {
+            map.insert(key.clone(), Arc::new(value()));
+        }
+        map.get(key).unwrap().clone()
+    }
+}
```

Comment:
I'd move this into a `mod append_only_map;` -- this is a nice data structure which isn't related to anything else in nearcore, so there's beauty in having just a small tiny isolated file for it.